### ISSUE TITLE
Suggestion: Introduce a Discussion tab

### ISSUE THREAD

**Issue by vjeux** [dear-github/dear-github#44]
**Problem with issues is to be forced to close them**
- As a maintainer, you are forced to take a binary decision, where in practice, the reality is a lot less black and white.
- The wording is very strong, your issue has been _closed_ which provokes a very negative reaction from the person on the other side. It makes dealing with issues a confrontation instead of a collaboration.
- Many issues we receive are not issues but questions, suggestions, thanks, reminders...

**Solution: introduce a discussion tab**
- Engineering wise, it would be very simple: the same backend as issues but without a badged count. (If you really insist, could badge for unread discussions).
- Many projects are moving discussions outside of github, for example for React we redirect people to Discourse. We'd rather stay on github since many discussions happen there anyway.
- It would reduce a lot of presure from maintainers as they are not forced to make a decision for all those discussions. The community can help each other out.

**Comment by origin1tech** [comment]
Couldn't agree with this more. For clarification the "confrontation" happens on nearly all projects large or small. The reality in development is you can make a case that no issue is truly closed. It is perhaps only closed perhaps for the currently know variables/conditions. Hence why you'll see the confrontation "hey guys this is still broke...here's the error I'm getting" comment. Could simply be a new set of params/variables not previously known this is just one example and why this merits discussion.

**Comment by stubailo** [comment]
Yeah, a while ago we decided to track feature requests on GitHub for Meteor, so now we have a very long list of things, some of which are bugs, some ideas, etc. It would be great to have a place to track things which aren't technically bugs, but are still nice to have on the repo.

Probably the biggest thing for me is that a large number of "issues" makes the project look like a lot of things are broken, when in reality some of those may be minor, or not even bugs at all.

**Comment by gaearon** [comment]
Yep. And a "Move to Discussions" button on issues, and vice versa :-)
This would be so much better for big projects.

**Comment by eyston** [comment]
Do the features of Discourse work well?  If it was integrated with Github (accounts, linking to code / issues / snippets / etc) would that work?

**Comment by kenwheeler** [comment]
@gaearon Move to Discussions would be beautiful.

**Comment by nafg** [comment]
Why not use gitter? Lots of project use it, and it has great integration with github (it can show issues activity) and other things (like CI)

You can put a badge or other link to gitter in the top of your readme.

**Comment by dralletje** [comment]
@nafg the problem I am seeing with gitter is the fact you can't keep track of multiple conversations. Conversations in issues can easily last a couple of days, and be re-opened when necessary. That is way harder when you have one chat log.

**Comment by stubailo** [comment]
Meteor has a discourse forum, and that works great for temporal discussions, but not for persistent tracking of different feature requests or ideas. For example, it would be great to be able to track the most popular feature requests, tag them, etc.

**Comment by jokeyrhyme** [comment]
There's a difference between closing an Issue and locking it.

From my experience, many GitHub users understand this difference, too. I see many "closed" Issues still continue on with discussions in them for months / years.

I've also seen many issues stay "open" indefinitely, as well, typically when there isn't a support or code solution, and the community is still comparing approaches.

And finally, I've seen discussions within Issues snowball, pulling in all sorts of tangents and side-topics, and ideally these become locked, with references to new single-topic Issues.

Maybe we just need all new projects to automatically have a "discussion" tag, or the opposite "todo" tag or something?

**Comment by claudiordgz** [comment]
Prioritizing the issues would be amazing, and would allow some companies I've worked with to drop their scrum backlog management tool. We also need some tool for epics that span multiple issues.

**Comment by joshmanders** [comment]
Indeed @claudiordgz. I am in that same boat. I've been working on a slow side project building a scrum like interface into Github using milestones, labels, issues and pull requests along with a command line tool and hopefully a github.io powered kanban board.

**Comment by josephsavona** [comment]
So tempting to :+1: this ;-) This is definitely an, err, *issue* that we're facing on [facebook/relay](https://github.com/facebook/relay) - do we close issues early to keep the issues list focused, or do we keep lots of issues open to promote more discussion?

**Comment by claudiordgz** [comment]
omg @joshmanders your tool sounds like it would be of great help. Some things I do them using scripts, for example raise an issue with refactor alert when a file or function reach x or y lines. But it is a bit of a hassle having to sell modules, small functions and prevent copy paste code over and over again when the tools that dev use could provide also recommendations based on so many works on application architecture.

**Comment by joshmanders** [comment]
@claudiordgz I pushed up what I have right now, it's still a heavy work in progress, but I'm building them out as git sub commands. https://github.com/joshmanders/dotfiles/tree/master/bin

Sooner or later I will probably rip them out of my dotfiles and put them in a proper package.

**Comment by claudiordgz** [comment]
@joshmanders this is great work will definitely check it out.

**Comment by naholyr** [comment]
This is only truely useful if there is a "Move to Discussion/Issues" button, but even without this :+1:

**Comment by petetnt** [comment]
I would be all in on introducing a discussion panel, if only to get a way to keep the discussion in one place as opposed to having to navigate to Discourse or Google Groups or some random Mailing List or another forum. 

I don't have an issue myself with `discussions-go-to-issues-too` approach, but obviously it can be a problem with larger projects. IMO the biggest problem in general is the kinds of questions that would be better suited to something like StackOverflow, where the question is that `is this a good way of doing this` or more often `why isn't this working`. In the suggested scheme you could always just dump those to discussions too, but that might just make the `Discussions` tab a graveyard for shitty issues.

**Comment by jlevy** [comment]
:+1: Agree something like this is needed, especially for projects (including mine) that are actually content, not code. It's increasingly common to (ab)use issues for discussion or questions.

It might make sense further to make tabs customizable, just like labels. It'd be nice to have Issues, Discussion, and Questions, for example. In fact, taking this further, they probably can be the same thing;  tabs would just be specially identified labels that affect UI and possibly state flow.

**Comment by davismj** [comment]
+1

**Comment by erikras** [comment]
And why can't I :+1: or :heart: or :star: comments that I love, like @gaearon's "Move to Discussions" comment? And this post in general.

Also, such binary responses would reduce the need for all the `+1` comments. No offense, @davismj.

**Comment by tejasmanohar** [comment]
+1

**Comment by Daniel15** [comment]
@nafg Gitter is good for real-time communication, but not for conversations you want to archive and potentially refer back to later (ie. have a permalink to it). A discussion tab would be more like a forum system, where you have multiple topics ideally with threaded replies.

**Comment by bsheikh** [comment]
+1

**Comment by simonc** [comment]
+1

**Comment by hcarrasc** [comment]
+1

**Comment by Stargator** [comment]
:+1:

**Comment by ron-wolf** [comment]
```c
++votes;
```

**Comment by onebree** [comment]
**TL;DR:** +1 with threaded capabilties

I think this is a fantastic idea. Right now, BetterSpecs has one open issue per guideline and tagged with "thoughts". As more guidelines are added, the issues list will grow immensely. Now, Disqus and Discourse are options for threaded feedback, but both require one additional step -- Signing in! Disqus does not have a Github OAuth, and Discourse is typically a new web page.

A Discussions tab in Github would be great if it had threaded capabilities like the above two services. It will also stop the issues list from reaching so high, making the project look unmaintained.

**Comment by nafg** [comment]
A "dual" to this proposal: https://github.com/almost/hubbub ;)

**Comment by mkonicek** [comment]
:100:

**Comment by bjfish** [comment]
How about just one tool called "Threads" (or something) with different types available? 
Issues, Discussions, Questions, or Custom types.

Issues feels like a general purpose communication tool but a lot of people restrict to using it only for issues because of its name.

**Comment by onebree** [comment]
I agree that "Questions" would be a nice category, but a lot of repos use Google Groups, a forum, or even Gitter for this. IMO I do not see Github becoming an all-in-one tool like Jira/Atlassian anytime soon. But it WOULD make everyone's lives easier sticking to a single service/platform for these things.

**Comment by jokeyrhyme** [comment]
:-1: 

I think this is a suggestion in search of a problem. It is not necessary to clutter the GitHub interface with every nuanced synonym in the English dictionary.

We're already getting commercial-quality code hosting for free. Are we expecting GitHub to host Discourse for free, too? Because I feel this suggestion is a slippery slope, with an endless chain of feature requests leading to "just host Discourse for free".

**Comment by fommil** [comment]
Getting what for free? I'm a paying customer.

**Comment by jokeyrhyme** [comment]
> Getting what for free? I'm a paying customer.

Then you can pay to either host Discourse yourself or otherwise, and get the best implementation of everything that has been discussed here. Easy.

**Comment by dylanaraps** [comment]
:+1:

**Comment by jokeyrhyme** [comment]
FYI: http://blog.discourse.org/2016/03/free-discourse-forum-hosting-for-community-friendly-github-projects/

> Do you run a popular open source project on GitHub? Then you might be eligible for a free, hosted Discourse. Your users can log in with GitHub logins and write posts in Markdown, and we’ve got tighter integration planned down the road, too.

**Comment by ron-wolf** [comment]
I don’t see why version control and discussion forums couldn’t be integrated by a single client.

**Comment by ericR-123** [comment]
+1 le upvote

**Comment by mitar** [comment]
Maybe the solution could be simply to not have only "open" or "closed" state of an issue, but also some other states. Like "in discussion". I do not think a new tab is really necessary.

And don't labels handle many of such cases as well?

**Comment by grahamperrin** [comment]
[2016-01-16](https://github.com/dear-github/dear-github/issues/44#issuecomment-172173662): 

> Gitter is … not for conversations you want to archive and potentially refer back to later (ie. have a permalink to it). … 

Archiving is automated. Each 'burst' of messages automatically gains a permalink. 

Recent https://github.com/gitterHQ/gitter/issues/1675 is _Permalink for every message in burst_ (was: _Allow a permalink without waiting five minutes between messages_).

> A discussion tab would be more like a forum system, where you have multiple topics ideally with threaded replies.

I do love interfaces that allow multiple threads – within a single topic – to be read, and understood, with ease. (That is, when it's not beneficial for a thread to be branched/split/spun off  into a separate topic.) 

For me, _traditionally_, part of that ease included **visual nesting** of some sort – for example, within the contraints of a Pipermail interface: 

![the freebsd-current august 2010 archive by thread -2-](https://cloud.githubusercontent.com/assets/192271/25272005/0538d596-267e-11e7-9a7c-d17774520ae9.png)

– there's also the benefit of a **thoughtful subject line** when posting. 

More recently, I find that ease of reading (and understanding) can be achieved **without** visual nesting. The key to this ease is **a smart interface** to discussions and for what it's worth, I treat Discourse as smartest of all. 

[2016-02-08](https://github.com/dear-github/dear-github/issues/44#issuecomment-181371385): 

> … BetterSpecs … 

Is that [Better Specs { rspec guidelines with ruby }](http://betterspecs.org/http://betterspecs.org/)?

[2016-02-12](https://github.com/dear-github/dear-github/issues/44#issuecomment-183479525): 

> I do not see Github becoming an all-in-one tool like Jira/Atlassian anytime soon. 

Agreed: I don't foresee GitHub becoming an all-in-one tool. 

(I used JIRA years ago, not enough to know its capabilities.) 

> … it WOULD make everyone's lives easier sticking to a single service/platform for these things. …

I disagree because GitHub is essentially, primarily, for audiences such as _developers_. That's a subset; a superset of audiences might include _testers_ and _end users_. I don't expect the essence of GitHub to change.

**Comment by mitar** [comment]
If discussions are introduced, I think it would be great if you could migrate issues to discussions and vice-versa.

**Comment by steveluo** [comment]
I have built a discussion service for GitHub, please take a look at the site https://gitq.com and some features https://gitq.com/GitQLab/gitq.com

Below is what a project looks like:

![screenshot](https://user-images.githubusercontent.com/467086/28715603-58749534-73cb-11e7-9db4-65e37e677cea.png)

Any suggestions are welcome!

**Comment by a472** [comment]
hello r,
I am a 2nd-year student of IT branch basically, I want to know how to use GitHub I do competitive coding on the various platform but once I heard about GitHub then I googled everything but I always confused and finally I give up. so, for which purpose people used this GitHub please help me ...

**Comment by mayou36** [comment]
@a472 this is the completely wrong place to ask such a question. Google "github" or better "git", or more general VCS (Version Control System). In short, it helps people to work on the same piece of code, share it and maintain it. And as an IT guy, I would highly recommend to look at it.

But please do not ask again such a question in a thread which is clearly (see title) for different purposes and do not answer either. Just google, use StackExchange and you will find a lot of satisfying answers ;)

**Comment by vesche** [comment]
A discussion tab would be highly beneficial to GitHub. When I [post a project on Reddit](https://www.reddit.com/r/netsec/comments/69torc/scanless_commandline_utility_for_using_public/), I get some nice dialogue that I would prefer to happen within [my project on GitHub](https://github.com/vesche/scanless/issues). And then I end up [getting feedback](https://github.com/vesche/scanless/issues/4) through an issue tracker that really should belong in a discussion. Also, sometimes users who made suggestions outside of GitHub might turn around and make pull requests and it's hard to correlate who did what.

+1 to a GitHub discussion tab! I don't think it would clutter the otherwise minimal interface.

**Comment by JCMais** [comment]
https://github.com/blog/2471-introducing-team-discussions almost that

**Comment by kaspersorensen** [comment]
+1

We could really use this in the [datacleaner/DataCleaner](https://gitq.com/datacleaner/DataCleaner) project since our current self-hosted discussion forum is going away (corporate decision to fork community and commercial project - and the forum is becoming a "casualty" in that forking) and we're trying to find an alternative that won't cause too much of a drop-off in the community because people will not want to create accounts more than one place. We're focusing on GitHub accounts because that allows people to get access to issues, PRs, wiki and all that sort of stuff. But discussions are ... missing.

**Comment by Daniel15** [comment]
@kaspersorensen

> We're focusing on GitHub accounts

Most forum systems these days will allow logging in using a GitHub account, avoiding the need to manually create an extra account.

**Comment by titou10titou10** [comment]
+1

**Comment by girng** [comment]
would benefit bigly (see reference) if feature came to fruition, +1

**Comment by jcyrss** [comment]
+1

**Comment by SMotaal** [comment]
Maybe a good starting point is acknowledge that "issues" is a strong word to begin with, because aside from GitHub (and mostly programmers), the rest of the world associate the word issues with some of the following:

- when you have issues, you see a therapist
- if you have an issue, file a complaint

So, if we only have issues, it just don't work for me. But if we had a better name for that, which I have pondered upon endlessly when looking at a boring issue, or one that is not actually an issue, we can use that. 

Obviously comes down to choice, because while some repos may never ever have issues in their allowed vocabulary, by all means some repos (probably the same) are destined only to dealing with issues.

It naturally follows that it seems reasonable for GitHub to experimentally allow repos to set their tab name, maybe from list to avoid the potential fiascos of collaborators exiting due to many issues of their own.

Now I can finally update my README to:
> If you have an issue with that, we suggest you find a repo with "issues" and open an issue there ❤️

**Comment by IrisDroidology** [comment]
I think that it would also be really good to add a discussion tab, with the same features & filters as issues, but categorised similar to a forum

**Comment by tommyalatalo** [comment]
I can't believe this is still not already in place, it would be incredibly useful to have a discussion forum for each project available right on github instead of having to search for obscure chatrooms or message boards. 

It would like many already mentioned also allow users to help eachother which beginner questions that may arise due to lack of experience with the project, instead of necessarily bothering the devs.

Please make this happen!

**Comment by jcyrss** [comment]
It's time for Microsoft to show some spirits :)

**Comment by SMotaal** [comment]
To be fair, it takes very little work, there are discussions already in orgs! And, to be fair, there are discussions already in orgs… ie for some, this `issue` would be `closed` yet for those asking it needs to be `reopened` — how GitHub could really use having a GitHub repo with the issues tab :)

**Comment by atymic** [comment]
Looks like this is in beta now!
https://github.com/zeit/next.js/discussions

Hopefully GA soon!

HN thread with some answers from the GitHub CEO:
https://news.ycombinator.com/item?id=22388639

**Comment by SMotaal** [comment]
I continue to want to urge the community to take a moment and consider… how selective feedback basically means features are optimized to certain conventions!

I get it, a lot of important people need to do their work without noise of those who are best to only learn to follow!

But honestly, unless they select those who are "open about their accessibility obstacles", all this progress is fast-tracking a majority, leaving everyone else behind, and invisible… systematically.

Others here please do me a favour before you misunderstand what I am saying and react out of some misplaced but equally hurtful sense of urgency.

If we cannot all be party to the discussion, and please don't say oh that's why we `aria`, those who are in those discussions will continue to become homogeneously exclusive, and will not have a clue that the rest are even real.

**Comment by CliffyE1968** [comment]
wow, four years after this was opened and still no discussion tab.

I'm pretty new to github and have been ordered to use it for a group based school project. While github excels at most stuff, it falls flat on its face when the group wants to have an online discussion.

I've found freedcamp.com has a very simple built in discussion system and also kaban style tasks. We could use freedcamp but it would be fantastic to just have everything we need for full collaboration in one handy place.

I'm pretty sure we will ditch github for everything except it's core functionality (code management) which is a shame. I would hope microsoft could whack a quick forum / discussion system into github soon :)

**Comment by Gizmotronn** [comment]
There is something in the beta coming soon :)

**Comment by Sunbreak** [comment]
> There is something in the beta coming soon :)

Looking forward to it. There're too many discussions than real issures in Flutter https://github.com/flutter/flutter/issues

**Comment by Gizmotronn** [comment]
@sunbreak do you think there will be labels/categories? Currently there aren't any on the beta...

I know that it's a beta, but it seems like something that should have been there from the start of the beta

**Comment by verdverm** [comment]
They seem to be rolling something out: https://github.com/cuelang/cue/discussions

**Comment by Sunbreak** [comment]
Vuetify got one: https://github.com/vuetifyjs/vuetify/discussions

**Comment by skywind3000** [comment]
Github discussions is fantastic !!

Can we have discussion APIs to manipulate the discussions' data ?

This can allow me to design my new UI for my project's discussion tab with only static html pages.

**Comment by vesche** [comment]
SO happy to see this being rolled out! I love it, add it site-wide please :rocket:

**Comment by Kubo2** [comment]
@GitHub What about allowing us to select which labels are used to count open issues in the badge?

**Comment by tomrav** [comment]
Is there anywhere to opt in to this feature? We'd be glad to use it in [Stylable](https://github.com/wix/stylable)

**Comment by verdverm** [comment]
Would also love to have an option to have them start like issues or PRs, so one could just discuss, or have a conversation around code as well, so like a PR where there is a branch-to-branch with a diff

**Comment by lostdesign** [comment]
Also here asking if there is any chance of opting in? I am running a 12k member community over on discord and would love to shift main question asking/helping onto GitHub discussions, using it as a main board for the community. Well, at least try it first if it works out at all.

**Comment by Sunbreak** [comment]
> > There is something in the beta coming soon :)
> 
> Looking forward to it. There're too many discussions than real issures in Flutter https://github.com/flutter/flutter/issues

Any update?

**Comment by steveluo** [comment]
I built a forum service for github before: https://gitq.com and an example forum looks like below:

![image](https://user-images.githubusercontent.com/467086/87525175-b836bb00-c6bb-11ea-9349-9e4855e7087f.png)

but it seems that people don't need a forum outside of the Github. what do you think?

**Comment by pontakornth** [comment]
There is a discussion tab in [snowpack](https://github.com/pikapkg/snowpack/discussion) repo now.

**Comment by makerjackie** [comment]
Now you can open "Discussion" for your repo in repo setting! Even it's only for beta now

**Comment by Davekibh** [comment]
Can we open the discussion tab cause I heard its only for beta.

**Comment by makerjackie** [comment]
> Can we open the discussion tab cause I heard its only for beta.

sure you can!

**Comment by Davekibh** [comment]
> > Can we open the discussion tab cause I heard its only for beta.
> 
> sure you can!

Okay Thank you!

**Comment by Sunbreak** [comment]
What about open Discussions for current repo, i.e. https://github.com/dear-github/dear-github?

**Comment by dreua** [comment]
@vjeux @dear-github I think this is resolved, can someone close this issue?

**Comment by vjeux** [comment]
It’s there!
